CustomVariable: support values with multiple properties (json values format)

grafakus opened the pull request:
**⚠️ Please wait for https://github.com/grafana/scenes/pull/1323 and update the Scenes in this PR before merging! ⚠️**

**What is this feature?**

Is built on top of https://github.com/grafana/scenes/pull/1236 and https://github.com/grafana/scenes/pull/1323, behind the `multiPropsVariables` feature toggle.

**Non-dynamic dashboards:**
<img width="3002" height="1356" alt="image" src="https://github.com/user-attachments/assets/c11aea73-1e72-4bd1-9341-aabce7b92fe8" />

**Dynamic dashboards:**
<img width="766" height="476" alt="image" src="https://github.com/user-attachments/assets/518d6267-55ea-4b27-92a1-090ae959cc29" />

**Preview, with deeply nested properties:**
<img width="765" height="821" alt="image" src="https://github.com/user-attachments/assets/a03fa981-e861-434c-b2fc-f7b5b0538ef3" />

**Autocomplete, with deeply nested properties:**
<img width="483" height="381" alt="image" src="https://github.com/user-attachments/assets/a8f49964-4b92-4d12-8f20-8c9b3aca7797" />

**Rollback warning:**
<img width="766" height="827" alt="image" src="https://github.com/user-attachments/assets/cf280be7-dd63-4836-9170-ff130aadae63" />

**Why do we need this feature?**

Feature [requested by customers](https://github.com/grafana/grafana-org/issues/555)

**Who is this feature for?**

All users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-org/issues/716

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle (set `multiPropsVariables = true` in `custom.ini`).
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

grafakus committed:
feat: WiP

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

grafakus committed:
Update v2 schema + gen types

grafakus committed:
chore: ...

bfmatei commented on apps/dashboard/pkg/apis/dashboard_manifest.go:
```diff
@@ -21,6 +21,8 @@ import (
 	v2beta1 "github.com/grafana/grafana/apps/dashboard/pkg/apis/dashboard/v2beta1"
 )
 
+var ()
```
This should be reverted manually.

bfmatei commented on public/app/features/dashboard-scene/settings/variables/components/QueryVariableForm.tsx:
```diff
@@ -102,48 +112,70 @@ export function QueryVariableEditorForm({
+            label="Enable access to all the fields of the query results"
+            description={
+              // TODO: add translation
+              <Trans i18nKey="">
```
There's a certain syntax on how to handle this. Also I'd use `t` instead of `<Trans />` here as it's a static string.

bfmatei reviewed (commented):
I left some comments around.

grafakus committed:
Strengthen valuesFormat type + cleanup generated files

grafakus committed:
Small preview fix when "All" option is checked

grafakus committed:
chore: Add comment

grafakus committed:
Translate CustomVariableEditor + improve JSON validation

grafakus committed:
VariableEditorForm checks to display preview with multiple props

grafakus committed:
Add some unit tests

grafakus committed:
Fix K8s Codegen Check

grafakus committed:
chore: Update to new Scenes version

grafakus committed:
Fix conflicts with main

grafakus committed:
feat: Update dynamic dashboards editors

grafakus committed:
Fix: hide options when multi properties exist on every var options

grafakus committed:
test(CustomVariableEditor): Add unit tests

grafakus committed:
chore: Better naming and minor test improvement

grafakus committed:
Remove temp switch in QueryVariableEditor - rely on options instead to determine if the variable has multi props

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

# Conflicts:
#	public/app/features/dashboard-scene/serialization/sceneVariablesSetToVariables.ts
#	public/app/features/dashboard-scene/serialization/transformSaveModelSchemaV2ToScene.ts

grafakus committed:
Add support for autocomplete on value properties

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

# Conflicts:
#	package.json
#	public/app/features/dashboard-scene/settings/variables/VariableEditorForm.tsx
#	public/app/features/dashboard-scene/settings/variables/components/QueryVariableForm.tsx
#	public/app/features/dashboard-scene/settings/variables/editors/CustomVariableEditor/ModalEditor.tsx
#	public/app/features/dashboard-scene/settings/variables/editors/QueryVariableEditor.tsx
#	yarn.lock

grafakus committed:
Simplify i18n

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

# Conflicts:
#	public/app/features/dashboard-scene/settings/variables/editors/CustomVariableEditor/CustomVariableEditor.tsx

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

grafakus committed:
Add count in preview

grafakus committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

grafakus committed:
Gate behind feature toggles

grafakus committed:
Fix build issues

yaelleC commented:
/deploy-to-hg

grafakus committed:
Fix preview of values: flatten properties to prevent runtime error

grafakus committed:
Memoize

grafakus committed:
Update schema to fix Go tests

grafakus committed:
Autocomplete support for deeply nested properties

yesoreyeram commented:
@grafakus - Sorry for not being clear. Find below the dashboard JSON without infinity element.

## Issue 1 : sub prop with `:text` suffix works incorrectly

I understood that multi prop variables doesn't support `:value` or `:text` suffix. But when used by mistake/accident, the `:value` suffix works fine but the `:text` suffix incorrectly returns the root variable text. Instead it should return the sub property value. Here in the below screenshot, `${user.email:text}` should return `Sincere@april.biz` instead of  `Leanne Graham`

<img width="393" height="261" alt="image" src="https://github.com/user-attachments/assets/6d8a89d3-260d-4440-a190-8d19ded2a5b9" />

<img width="466" height="684" alt="image" src="https://github.com/user-attachments/assets/2bc33272-922e-45c1-a99e-3495e86c397b" />

## Issue 2 : Preview values not showing for props with `.` in the name

<img width="390" height="119" alt="image" src="https://github.com/user-attachments/assets/c7a06eda-7429-4e5e-9b02-1076d52d7877" />

<img width="1726" height="404" alt="image" src="https://github.com/user-attachments/assets/c906edd7-db4b-422e-85c1-39a7e906eb9d" />


<details>

<summary>Dashboard JSON</summary>

## Dashboard JSON

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "links": [],
  "panels": [
    {
      "fieldConfig": {
        "defaults": {},
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "code": {
          "language": "plaintext",
          "showLineNumbers": false,
          "showMiniMap": false
        },
        "content": "# Single variable \n\n| Template  | Interpolated Value |\n|-----------|-------------------|\n| $\\{user} | ${user} |\n| $\\{user:text} | ${user:text} |\n| $\\{user:value} | ${user:value} |\n| $\\{user.username} | ${user.username} |\n| $\\{user.username:value} | ${user.username:value} |\n| $\\{user.username:text} | ${user.username:text} |\n| $\\{user.user name} | ${user.user name} |\n| $\\{user.user name:value} | ${user.user name:value} |\n| $\\{user.user name:text} | ${user.user name:text} |\n| $\\{user.user.name} | ${user.user.name} |\n| $\\{user.user.name:value} | ${user.user.name:value} |\n| $\\{user.user.name:text} | ${user.user.name:text} |\n| $\\{user.email} | ${user.email} |\n| $\\{user.email:value} | ${user.email:value} |\n| $\\{user.email:text} | ${user.email:text} |",
        "mode": "markdown"
      },
      "pluginVersion": "12.4.0-pre",
      "title": "",
      "type": "text"
    }
  ],
  "preload": false,
  "schemaVersion": 42,
  "tags": [],
  "templating": {
    "list": [
      {
        "allowCustomValue": false,
        "current": {
          "text": "1",
          "value": "1"
        },
        "name": "user",
        "options": [],
        "query": "[\n    {\n        \"text\": \"Leanne Graham\",\n        \"value\": 1,\n        \"username\": \"Bret\",\n        \"email\": \"Sincere@april.biz\",\n        \"user name\": \"Bret\",\n        \"user.name\": \"Bret\"\n    },\n    {\n        \"text\": \"Ervin Howell\",\n        \"value\": 2,\n        \"username\": \"Antonette\",\n        \"email\": \"Shanna@melissa.tv\",\n        \"user name\": \"Antonette\",\n        \"user.name\": \"Antonette\"\n    },\n    {\n        \"text\": \"Clementine Bauch\",\n        \"value\": 3,\n        \"username\": \"Samantha\",\n        \"email\": \"Nathan@yesenia.net\",\n        \"user name\": \"Samantha\",\n        \"user.name\": \"Samantha\"\n    },\n    {\n        \"text\": \"Patricia Lebsack\",\n        \"value\": 4,\n        \"username\": \"Karianne\",\n        \"email\": \"Julianne.OConner@kory.org\",\n        \"user name\": \"Karianne\",\n        \"user.name\": \"Karianne\"\n    },\n    {\n        \"text\": \"Chelsey Dietrich\",\n        \"value\": 5,\n        \"username\": \"Kamren\",\n        \"email\": \"Lucio_Hettinger@annie.ca\",\n        \"user name\": \"Kamren\",\n        \"user.name\": \"Kamren\"\n    },\n    {\n        \"text\": \"Mrs. Dennis Schulist\",\n        \"value\": 6,\n        \"username\": \"Leopoldo_Corkery\",\n        \"email\": \"Karley_Dach@jasper.info\",\n        \"user name\": \"Leopoldo_Corkery\",\n        \"user.name\": \"Leopoldo_Corkery\"\n    },\n    {\n        \"text\": \"Kurtis Weissnat\",\n        \"value\": 7,\n        \"username\": \"Elwyn.Skiles\",\n        \"email\": \"Telly.Hoeger@billy.biz\",\n        \"user name\": \"Elwyn.Skiles\",\n        \"user.name\": \"Elwyn.Skiles\"\n    },\n    {\n        \"text\": \"Nicholas Runolfsdottir V\",\n        \"value\": 8,\n        \"username\": \"Maxime_Nienow\",\n        \"email\": \"Sherwood@rosamond.me\",\n        \"user name\": \"Maxime_Nienow\",\n        \"user.name\": \"Maxime_Nienow\"\n    },\n    {\n        \"text\": \"Glenna Reichert\",\n        \"value\": 9,\n        \"username\": \"Delphine\",\n        \"email\": \"Chaim_McDermott@dana.io\",\n        \"user name\": \"Delphine\",\n        \"user.name\": \"Delphine\"\n    },\n    {\n        \"text\": \"Clementina DuBuque\",\n        \"value\": 10,\n        \"username\": \"Moriah.Stanton\",\n        \"email\": \"Rey.Padberg@karina.biz\",\n        \"user name\": \"Moriah.Stanton\",\n        \"user.name\": \"Moriah.Stanton\"\n    }\n]",
        "type": "custom",
        "valuesFormat": "json"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "browser",
  "title": "New dashboard",
  "version": 0,
  "weekStart": ""
}
```

</details>

grafakus committed:
Fix import dashboard with CustomVariable and json valuesFormat

grafakus committed:
Fix go-fmt

grafakus committed:
More gens to fix build

grafakus committed:
Fix unit tests

grafakus committed:
Fix preview issue when property names have dots

grafakus commented:
@yesoreyeram thank you for testing and finding these issues!

> **Issue 1** : sub prop with `:text` suffix works incorrectly

Addressed in https://github.com/grafana/scenes/pull/1323

> **Issue 2** : Preview values not showing for props with `.` in the name

Fixed by https://github.com/grafana/grafana/pull/113844/changes/220701ece2c0d91e67693be5aec90285b465a27f

And by using the dashboard JSON you provided, I also found an issue with the import, which has been fixed in https://github.com/grafana/grafana/pull/113844/changes/0575d5725fc91712f86308bd8e6b43b4c4471045

grafakus committed:
Fix CI integration tests

idastambuk commented:
Did a similar test to Sriram's and it looks good, just one question:
Is it intended that ${custom:value} interpolates a multi value by leaving the brackets in?

<img width="500" height="567" alt="Screenshot 2025-12-30 at 14 38 01" src="https://github.com/user-attachments/assets/f8212d50-2dec-4bfe-be86-6f61779fa751" />

grafakus committed:
Improve preview: flatten deeply nested properties instead of stringifying them

idastambuk committed:
Display old label when toggle is disabled

idastambuk committed:
Add field to v1 schema

idastambuk committed:
add tracking for csv

idastambuk committed:
Update scenes

yaelleC reviewed (approved):
reviewed the last 3 commits:
[Display old label when toggle is disabled](https://github.com/grafana/grafana/pull/113844/commits/918c22e499486d784c2a16eff56eb29344fbcd43)
[Add field to v1 schema](https://github.com/grafana/grafana/pull/113844/commits/78f5d936a742e79d140ae3df48b9a820bd5e5363)
[add tracking for csv variables](https://github.com/grafana/grafana/pull/113844/commits/383773978c9f034718c56c1586d4c36806310e37)

LGTM 🚀

idastambuk committed:
Merge branch 'main' into grafakus/multi-dimensional-vars-ui

idastambuk committed:
add unit tests for tracking
